cpu/cortexm_common/include/systick.h

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (C) 2020 ML!PA Consulting GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    drivers_periph_systick SysTick Timer
+ * @ingroup     cpu_cortexm_common
+ * @ingroup     drivers_periph
+ * @brief       Low-level timer based on SysTick
+ *
+ * SysTick is a 24 bit down-counting timer that is implemented on every Cortex-M
+ * processor. It runs at the same frequency as the CPU and will generate an
+ * interrupt when it reaches zero.
+ *
+ * To comply with the RIOT timer interface that expects an up-counting, monotonic
+ * timer, SysTick is augmented by software to fulfill these requirements.
+ * This also adds a software prescaler to simulate arbitrary frequencies and introduces
+ * a virtual, 32 bit counter.
+ * That allows for longer intervals than would be possible with the SysTick hardware
+ * peripheral alone.
+ *
+ * @{
+ *
+ * @file
+ * @brief       Low-level SysTick timer peripheral driver interface definitions
+ *
+ * @author      Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ */
+
+#ifndef SYSTICK_H
+#define SYSTICK_H
+
+#include <stdint.h>
+#include <limits.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief   SysTick is extended to 32 bit by software
+ */
+#define SYSTICK_TIMER_MAX_VALUE UINT32_MAX
+
+/**
+ * @brief   Signature of event callback functions triggered from interrupts
+ *
+ * @param[in] arg       optional context for the callback
+ */
+typedef void (*systick_cb_t)(void *arg);
+
+/**
+ * @brief Initialize the SysTick timer
+ *
+ * The timer is configured in up-counting mode and will count until
+ * SYSTICK_TIMER_MAX_VALUE until overflowing.
+ *
+ * The timer will be started automatically after initialization with interrupts
+ * enabled.
+ *
+ * @param[in] freq          requested number of ticks per second
+ * @param[in] cb            this callback is called in interrupt context, the
+ *                          emitting channel is passed as argument
+ * @param[in] arg           argument to the callback
+ */
+void SysTick_timer_init(unsigned long freq, systick_cb_t cb, void *arg);
```

Comment:
No camel case, please.

Also: If there is at most a single SysTick timer, do we really need to use indirect calls? Couldn't we just declare a `void systick_callback(void)` function that needs to be provided by the user of this API? (Also: With at most a single timer of this time, no user argument is really needed.)

Reply:
What about the indirect calls? I'd say there is no need for a `systick_cb_t cb` argument, as at most one systick timer is present. We could just ask the user to implement `void systick_callback(void)` instead and use a direct call. That would free a bit of RAM for holding the address, and a bit of ROM for the indirect call.

A direct call can also help to keep the instruction pipeline filled, with notable performance improvements as result.

Reply:
I thought I'd keep the interface as similar as the periph timer interface and allow for run-time reconfiguration.
But looking at the timer interface, they only allow re-configuration on init().

And while it's possible to call `systick_timer_init` multiple times, that's not possible with all periph timers I think.

So yea, I'll add a fixed `systick_callback(void)`.